File: flashcard-api/src/main/java/com/flashcard/model/Deck.java
package com.flashcard.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import jakarta.validation.constraints.NotBlank;
import java.time.Instant;

/**
 * MongoDB document representing a Flashcard Deck (a collection of cards).
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "decks")
public class Deck {

    @Id
    private String id;

    @NotBlank(message = "Deck name is required")
    private String name;

    private String description;

    /**
     * Owner's user ID. Indexed for fast lookup of decks by user.
     */
    @Indexed
    private String userId;

    /**
     * Number of cards in this deck (denormalized for performance).
     */
    @Builder.Default
    private int cardCount = 0;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}

File: app/src/main/java/com/example/flashcardapp/domain/model/Deck.kt
package com.example.flashcardapp.domain.model

import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes

data class Deck(
    val id: String,
    val categoryId: String = "",
    val name: String,
    val description: String? = null,
    val themeColor: String? = null,
    @DrawableRes val iconResId: Int? = null,
    @ColorRes val backgroundResId: Int? = null,
    val isPublic: Boolean = false,
    val createdAt: String? = null,
    val updatedAt: String? = null,
    val cards: List<FlashCard> = emptyList(),
    val customCardCount: Int? = null
) {
    val cardCount: Int
        get() = customCardCount ?: cards.size
}
